server/src/main/java/io/unitycatalog/server/service/PermissionService.java

```diff
@@ -0,0 +1,278 @@
+      @Param("name") @AuthorizeKey(CATALOG) String name, UpdatePermissions request) {
+    return updateAuthorization(CATALOG, name, request);
+  }
+
+  @Patch("/schema/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #schema, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateSchemaAuthorization(
+      @Param("name") @AuthorizeKey(SCHEMA) String name, UpdatePermissions request) {
+    return updateAuthorization(SCHEMA, name, request);
+  }
+
+  @Patch("/table/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #table, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateTableAuthorization(
+      @Param("name") @AuthorizeKey(TABLE) String name, UpdatePermissions request) {
+    return updateAuthorization(TABLE, name, request);
+  }
+
+  @Patch("/function/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #function, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateFunctionAuthorization(
+      @Param("name") @AuthorizeKey(FUNCTION) String name, UpdatePermissions request) {
+    return updateAuthorization(FUNCTION, name, request);
+  }
+
+  @Patch("/volume/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #volume, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateVolumeAuthorization(
+      @Param("name") @AuthorizeKey(VOLUME) String name, UpdatePermissions request) {
+    return updateAuthorization(VOLUME, name, request);
+  }
+
+  @Patch("/registered-model/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #registered_model, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateRegisteredModelAuthorization(
+      @Param("name") @AuthorizeKey(REGISTERED_MODEL) String name, UpdatePermissions request) {
+    return updateAuthorization(REGISTERED_MODEL, name, request);
+  }
+
+  private HttpResponse updateAuthorization(
+      SecurableType securableType, String name, UpdatePermissions request) {
+    UUID resourceId = getResourceId(securableType, name);
+    List<PermissionsChange> changes = request.getChanges();
+    Set<UUID> principalIds = new HashSet<>();
+    changes.forEach(
+        change -> {
+          String principal = change.getPrincipal();
+          User user = USER_REPOSITORY.getUserByEmail(principal);
+          UUID principalId = UUID.fromString(Objects.requireNonNull(user.getId()));
+          principalIds.add(principalId);
+          change
+              .getAdd()
+              .forEach(
+                  privilege ->
+                      //  Privileges should always be a superset of Privilege so this _should_
+                      // always be non-null but let's be safe anyway.
+                      Optional.ofNullable(Privileges.fromPrivilege(privilege))
```

Comment:
Where is the validation done that the right privilege is being applied to the correct object (for example, USE_CATALOG isn't being applied to a schema)?

Reply:
This is missing. If this is a P0 for v0.2 I can probably add something quickly otherwise I would suggest deferring.

Reply:
Can you add a GH issue to track this for v0.3?

Reply:
Can you add a GH issue to track this?